[FEATURE][2] Search and Sort

mathiramilo-lit opened the pull request:
# Search and Sort - 2

## Description ⭐

I've implemented products searching and sorting.

### Type of change 🏃

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

## Screenshots 📷

## Checklist ✅

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

mathiramilo-lit committed:
refactor: move fetch logic to app

mathiramilo-lit committed:
refactor: update to react and ts best practices

mathiramilo-lit committed:
chore: add fonts and .vscode

mathiramilo-lit committed:
refactor: implement component composition and useProducts

mathiramilo-lit committed:
feat: add debounced search

mathiramilo-lit committed:
feat: add sort

mathiramilo-lit committed:
feat: add favorites

mathiramilo-lit committed:
feat: add url state handling

hyanez-lightit commented on src/components/button.tsx:
```diff
@@ -1,14 +1,17 @@
-  onClick: () => void;
-  loading?: boolean;
-}
+type ButtonProps = ComponentPropsWithoutRef<"button"> &
+  PropsWithChildren & {
+    onClick: MouseEventHandler<HTMLButtonElement>;
+    loading?: boolean;
+  };
```
button already has children & onclick 
```suggestion
type ButtonProps = ComponentPropsWithoutRef<"button"> & {
    loading?: boolean;
  };
```

mathiramilo-lit replied:
Thanks, you're right! I wanted to make the onClick prop mandatory for this component. What’s the best way to achieve that while keeping the existing props? I appreciate your guidance!

hyanez-lightit replied:
Then yes you can overwrite it as you did (just remove the PropsWithChildren because thats already there).
But, on that note, what if you want to use this button as the default submit button of a form? It doesn't need an onclick handler there, is it wise to overwrite HTML's native attributes?

mathiramilo-lit replied:
Absolutely not. I’ll remove PropsWithChildren and ensure the button supports use as a default submit button without overwriting native attributes. Thanks!

hyanez-lightit commented on src/components/sort-by.tsx:
```diff
@@ -29,22 +35,26 @@ export const SortBy = () => {
-        </button>
+        {options.map((option, index) => (
+          <button
+            key={index}
```
[Avoid using index as key](https://react.dev/learn/rendering-lists#why-does-react-need-keys)

danielaWaldeck95 replied:
Some more documentation about this on our [standards](https://lightit.slite.com/app/docs/LBfbP_38cP-oT2?noteModalId=9qRM-JqhgCpj1q)!

mathiramilo-lit replied:
Thank you for the feedback! I’m aware that using the index as a key is not ideal due to potential issues with reordering. In this case, the options array is static and only contains 4 objects, so I used the index as a quick solution. However, I agree that it's better to follow best practices, so I'll update the code by assigning a unique id to each object and use that as the key. Thanks again for the reminder!

hyanez-lightit replied:
you can use the option's label if its unique

hyanez-lightit commented on src/components/sort-by.tsx:
```diff
@@ -1,23 +1,29 @@
+import { Order, ProductField, SortOption } from '../types';
 
-export const SortBy = () => {
+interface SortByProps {
+  options: SortOption[];
+  setSort: (sortBy: ProductField, order: Order) => void;
+}
+
+export const SortBy = ({ options, setSort }: SortByProps) => {
```
this component definition & name feels off
also setSort could use an object-parameter instead of two positional ones

mathiramilo-lit replied:
I agree! I'll work on improving the component definition and name, as well as following your recommendation to use an object parameter for setSort. Thanks!

hyanez-lightit commented on src/hooks/use-debounce.ts:
```diff
@@ -0,0 +1,17 @@
+  const [debouncedValue, setDebouncedValue] = useState(value);
+
+  useEffect(() => {
+    const handler = setTimeout(() => {
+      setDebouncedValue(value);
+    }, delay);
+
+    return () => {
+      clearTimeout(handler);
+    };
```
this is not particularly the hanlder but the timeout id
```suggestion
    const timeoutId = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(timeoutId);
    };
```

mathiramilo-lit replied:
Thanks! You're right, I'll rename it.

hyanez-lightit commented on src/hooks/use-products.ts:
```diff
@@ -24,7 +29,7 @@ export const Products = () => {
 
     try {
       const res = await axios.get(
-        `https://dummyjson.com/products?limit=${LIMIT}&skip=${pagination.skip}`,
+        `https://dummyjson.com/products${query ? `/search?q=${query}&` : '?'}limit=${LIMIT}&skip=${pagination.skip}${sort ? `&sortBy=${sort.sortBy}&order=${sort.order}` : ''}`,
       );
```
[Wheel is already invented ](https://axios-http.com/docs/req_config):) (adjust accordingly)
```
const res = await axios.get(
        'https://dummyjson.com/products',
        {  params: { q: query, limit: LIMIT, skip: pagination.skip, ...sort } }
      );
```

mathiramilo-lit replied:
I was feeling something weird about this. That’s a great piece of code suggestion, thanks for the snippet! I’ll update the request to use the params object.

mathiramilo-lit committed:
chore: add import alias

Update tsconfig files and vite.config.ts to add import alias, and create cspell.json file to add unrecognized words.

mathiramilo-lit committed:
fix: pr comments

* update axios calls to use params
* change "debounce" to "timeoutId" in useDebounce
* change SortBy component: rename to "SortDropdown", object-parameter to setSort and remove useEffect
* add ? for possible undefined values
